### ISSUE TITLE
Subclasses of frozen class don't respect frozen=False

### ISSUE THREAD

**Issue by Vegemash** [microsoft/pyright#4193]
**Describe the bug**
When a class is declared as `frozen=False` but inherits from something defined as `frozen=True`, the class is mutable and will not produce errors at runtime. Pyright seems to see the child class as still frozen and throws an error

**To Reproduce**
Run pyright on the code below

**Expected behavior**
No error on subclasses that define `frozen=False`

**Screenshots or Code**
```python
from pydantic import BaseModel, Field


class T(BaseModel, frozen=True):
    a: str = "Eric"


class B(T, frozen=False):
    b: str = "Traut"

class C(T):
    c: str = Field(default="", frozen=False)

class D(T, frozen=False):
    d: str = Field(default="", frozen=False)

b_instance = B() : B
b_instance.b = "🐟"   # This results in a pyright error but no runtime error

assert b_instance.b == "🐟"
assert T.__config__.frozen is True
assert B.__config__.frozen is False
D().d = ""  # pyright error but no runtime error
C().c = ""  # Fails at runtime but no pyright error
```
pyright 1.1.279
pydantic 1.10.2

**VS Code extension or command-line**
Command line

**Additional context**
I don't know if this should produce an error at runtime or not, I couldn't figure an answer from PEP681. Also adding `Field(frozen=False)` to the fields you want to be able to modify seems to have inconsistent results.

**Comment by erictraut** [comment]
There is a bug here, but it's not the one you are reporting.

Both `dataclass` and `attrs` (which is similar to dataclass) disallow a frozen class to derive from a non-frozen class and vice versa. The reasons for this are explained in [this bpo](https://bugs.python.org/issue43176).

Pyright correctly emits an error when a frozen dataclass derives from a non-frozen dataclass, but it fails to emit an error when the converse occurs. Both of these situations generate runtime exceptions, so this is a false negative. I've just updated the logic to address this oversight. This will be fixed in the next release.

Pydantic uses PEP 681 (Dataclass Transforms) which means that type checkers assume that it has dataclass semantics. Pyright is enforcing those semantics here. Based on what you're saying, pydantic deviates from the standard dataclass behavior in that it (unsafely) allows non-frozen classes to derive from a frozen class and vice versa. While pydantic may allow this at runtime, it's not generally safe from a type perspective, and combining frozen and non-frozen dataclass-like classes in a class hierarchy is disallowed by PEP 681.

**Comment by erictraut** [comment]
By the way, I appreciate the use of the 🐟 emoji. Nice touch. 😀
